### PR TITLE
feat: support f32 attention output in FA2 template

### PR DESCRIPTION
**Pull request by yzh119**
For bf16 kernels, we need to use f32 as the intermediate data type for split-k partial output to avoid numerical errors. This PR adds support for f32 output in CUDA templates.

This PR is the first step towards addressing the bf16 kernel numerical issues.

The remaining tasks include:
1. do not pre-apply `sm_scale` to query, apply `sm_scale` to logits instead.
2. change the split-k default partial output dtype to float32.

### CODE REVIEW

**Comment by zhyncs** [comment]
QQ How about this result? https://github.com/flashinfer-ai/flashinfer/blob/824ce40a48dda12ec600c9fa8f8b96246cf13b85/tests/test_hopper.py#L124

**Comment by zhyncs** [comment]
Ok I'll delete the tmp solution branch https://github.com/flashinfer-ai/flashinfer/tree/bf16-try

**Comment by yzh119** [comment]
This PR only modifies the C++ template, there will be 3 PRs in total.
